libraries/plugins/fork_info/fork_info_plugin.cpp

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (c) 2015 Cryptonomex, Inc., and contributors.
+ *
+ * The MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <steemit/fork_info/fork_info_plugin.hpp>
+
+#include <steemit/protocol/types.hpp>
+
+#include <steemit/chain/database.hpp>
+#include <steemit/app/api.hpp>
+#include <steemit/app/database_api.hpp>
+
+#include <fc/network/http/websocket.hpp>
+#include <fc/rpc/websocket_api.hpp>
+#include <fc/api.hpp>
+#include <fc/smart_ref_impl.hpp>
+
+namespace steemit { namespace plugin { namespace fork_info {
+
+namespace detail {
+   using namespace steemit::protocol;
+
+   class fork_info_plugin_impl
+   {
+      public:
+         fork_info_plugin_impl( fork_info_plugin& plugin )
+            :_self( plugin ) {}
+         virtual ~fork_info_plugin_impl() {}
+
+         void on_block( const signed_block& b );
+
+         fork_info_plugin&       _self;
+         std::deque<signed_block> _last_blocks;
+   };
+   void fork_info_plugin_impl::on_block( const signed_block& b )
```

Comment:
This gets called every block that is pushed. So when we switch forks, all the blocks on the other fork will be pushed. They will have already been pushed at some point. It would be possible to have the vector have the blocks in it in the following order.

{1a, 1b, 1a, 2a, 1b, 2b, 1a, 2a, 3a, 1b, 2b, 3b, ... }

The 21 block limit could be exceeded before the block is potentially made reversible and repeat blocks may be logged.

Instead, I would use a map from block num to a vector of blocks with that number. Then, when a the last irreversible block number is increased, check blocks with num less than or equal to LIB until there is no vector of blocks. For each vector check if each block is in the current chain. If it is not, then log it.